Source/Events.Store.MongoDB/Streams/Filters/FilterDefinitionDiscriminatorConvention.cs

```diff
@@ -0,0 +1,133 @@
+        /// </summary>
+        public enum FilterType
+        {
+            /// <summary>
+            /// Represents filters that get filtered in the head.
+            /// </summary>
+            Remote,
+
+            /// <summary>
+            /// Represents filters that get filtered in the runtime depending on a given declaration.
+            /// </summary>
+            EventTypeId
+        }
+
+        /// <summary>
+        /// Gets the fields name in mongo.
+        /// </summary>
+        public string ElementName => Type;
+
+        /// <summary>
+        /// Gets the value to be put into the ElementName field. The enum is saved as a string so that it looks prettier
+        /// when inspecting the db.
+        /// </summary>
+        /// <param name="nominalType">The nominal type.</param>
+        /// <param name="actualType">The actual type.</param>
+        /// <returns>A string.</returns>
+        public BsonValue GetDiscriminator(Type nominalType, Type actualType)
+        {
+            if (actualType == typeof(FilterDefinition) || actualType == typeof(PublicFilterDefinition))
+            {
+                return nameof(FilterType.Remote);
+            }
+            else if (actualType == typeof(TypePartitionFilterDefinition))
+            {
+                return nameof(FilterType.EventTypeId);
+            }
+            else
+            {
+                throw new UnsupportedTypeForFilterDefinitionDiscriminatorConvention(actualType);
+            }
+        }
+
+        /// <summary>
+        /// Gets the correct type when deserializing objects from  <see cref="AbstractFilterDefinition"/> collection.
+        /// </summary>
+        /// <param name="bsonReader">A <see cref="IBsonReader"/>.</param>
+        /// <param name="nominalType">The nominal type.</param>
+        /// <returns>The actual wanted type.</returns>
+        public Type GetActualType(IBsonReader bsonReader, Type nominalType)
+        {
+            ThrowIfNominalTypeIsIncorrect(nominalType);
+            var bookmark = bsonReader.GetBookmark();
+            bsonReader.ReadStartDocument();
+            Guid id = default;
+            FilterType filterType = default;
+            bool isFilterSet = false;
+            bool publicValue = default;
+            bool isPublicSet = false;
+
+            while (bsonReader.ReadBsonType() != BsonType.EndOfDocument)
+            {
+                switch (bsonReader.ReadName())
+                {
+                    case Type:
+                        filterType = Enum.Parse<FilterType>(bsonReader.ReadString());
+                        isFilterSet = true;
+                        break;
+                    case Public:
+                        publicValue = bsonReader.ReadBoolean();
+                        isPublicSet = true;
+                        break;
+                    case "_id":
+                        id = (Guid)bsonReader.ReadBinaryData();
```

Comment:
This seems a bit dangerous, as there is special things setup in the MongoClient for how to deal with `Guid`s.

Reply:
This method is probably safer - passed along with the same GuidRepresentation that we use in the client (which should be the standard one).
https://mongodb.github.io/mongo-csharp-driver/2.5/apidocs/html/M_MongoDB_Bson_BsonBinaryData_ToGuid_1.htm